qutebrowser/browser/history.py

```diff
@@ -184,16 +187,18 @@ def __init__(self, progress, parent=None):
             self.metainfo.try_recover()
             rebuild_completion = self.metainfo['force_rebuild']
 
-        if sql.user_version_changed():
-            # If the DB user version changed, run a full cleanup and rebuild the
-            # completion history.
-            #
-            # In the future, this could be improved to only be done when actually needed
-            # - but version changes happen very infrequently, rebuilding everything
-            # gives us less corner-cases to deal with, and we can run a VACUUM to make
-            # things smaller.
-            self._cleanup_history()
-            rebuild_completion = True
+        if self._database.user_version_changed():
+            with self._database.transaction():
+                # If the DB user version changed, run a full cleanup and rebuild the
+                # completion history.
+                #
+                # In the future, this could be improved to only be done when actually
+                # needed - but version changes happen very infrequently, rebuilding
+                # everything gives us less corner-cases to deal with, and we can run a
+                # VACUUM to make things smaller.
+                self._cleanup_history()
+                rebuild_completion = True
+                self._database.upgrade_user_version()
```

Comment:
I suppose this should only be called after the completion is rebuilt, though?

Reply:
I wanted to do this but:
1. I cannot call `VACUUM` inside a transaction, which `_rebuild_completion` does. I had also moved the logic in `Database.__init__` that sets `PRAGMA journal_mode` to `Database.upgrade_user_version` but that also cannot be called inside a transaction, so I had to move it back.
2. The completion can be rebuilt independently of `user_version` changing, and even has its own logic to retry if it fails or was interrupted (`CompletionMetaInfo.force_rebuild`).